libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -212,25 +221,94 @@ errorHandlers =
     ]
 {-# INLINE errorHandlers #-}
 
+-- | If the log level is less sensitive than 'Debug' just call the underlying app unchanged.
+-- Otherwise, pull a copy of the request body before running it, and if response status is @>=
+-- 400@, log the entire request, including the body.
+--
+-- The request sanitizer is called on the 'Request' and its body before it is being logged,
+-- giving you a chance to erase any confidential information.  Eg., if you're only interested
+-- in one end-point that has no sensitive information, use 'onlyLogEndpoint'.
+--
+-- WARNINGS:
+--
+--  * This may log confidential information if contained in the request.  Use the sanitizer to
+--    avoid that.
+--  * This does not catch any exceptions in the underlying app, so consider calling
+--    'catchErrors' before this.
+--  * Be careful with trying this in production: this puts a performance penalty on every
+--    request (unless level is less sensitive than 'Debug').
+heavyDebugLogging
+    :: ((Request, LByteString) -> Maybe (Request, LByteString))
+    -> Level -> Logger -> Middleware
+heavyDebugLogging sanitizeReq lvl lgr app = \req cont -> do
+    (bdy, req') <- if lvl <= Debug
+        then cloneBody req
+        else pure ("n/a", req)
+    app req' $ \resp -> do
+        forM_ (sanitizeReq (req', bdy)) $ \(req'', bdy') ->
+            when (statusCode (responseStatus resp) >= 400) $ logBody req'' bdy'
+        cont resp
+  where
+    cloneBody :: Request -> IO (LByteString, Request)
+    cloneBody req = do
+        bdy <- lazyRequestBody req
+        requestBody' <- emitLByteString bdy
+        pure (bdy, req { requestBody = requestBody' })
+
+    logBody :: Request -> LByteString -> IO ()
+    logBody req bdy = Log.debug lgr logMsg
+      where
+        logMsg = field "request" (fromMaybe "N/A" $ lookupRequestId req)
+               . field "request_details" (show req)
+               . field "request_body" bdy
+               . msg (val "full request details")
+
+onlyLogEndpoint :: Method -> [Text] -> (Request, LByteString) -> Maybe (Request, LByteString)
+onlyLogEndpoint mth pathi out@(req, _) =
+    if requestMethod req == mth && pathInfo req == pathi
+    then Just out
+    else Nothing
+
+emitLByteString :: LByteString -> IO (IO ByteString)
+emitLByteString lbs = do
```

Comment:
This is pretty wild stuff; What does it gain us? We've already forced the bytestring into memory (and consumed it from the client) so why are we trying to re-chunk it again? Could we just use `pure` on the forced bytestring instead?

I'd strongly prefer to avoid this as a premature optimization unless it's absolutely required for other reasons.

Reply:
no you haven't tested it, it doesn't compile.  :-)

but i'll change it, and leave the "wild" implementation in the haddocks.

Reply:
Haha; I tested the implementation, not my fault if you don't have the right things in scope 😋